contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+
+            let Route { hops, to } = from_binary(&msg)?;
+
+            if hops.len() < 2 {
+                return Err(StdError::generic_err("route must be at least 2 hops"));
+            }
+
+            for i in 1..(hops.len() - 1) {
+                if hops[i].from_token.address == None
+                    || hops[i].from_token.code_hash == None
+                    || hops[i].from_token.native_denom != None
+                {
+                    return Err(StdError::generic_err(
+                        "cannot route via uscrt. uscrt can only be route input token or output token.",
+                    ));
+                }
+            }
+
+            let first_hop: Hop = hops[0].clone();
+
+            let received_from_token_of_first_hop =
+                Some(env.message.sender) == first_hop.from_token.address;
+
+            let mut received_scrt = false;
+            let mut scrt_is_first_hop = false;
+            if env.message.sent_funds.len() == 1 {
+                received_scrt = env.message.sent_funds[0].amount == amount
+                    && env.message.sent_funds[0].denom == "uscrt";
+                scrt_is_first_hop = first_hop.from_token.native_denom == Some("uscrt".into());
+            }
+
+            if !received_from_token_of_first_hop && !(received_scrt && scrt_is_first_hop) {
+                return Err(StdError::generic_err(
+                    "route can only be initiated by sending here the token of the first hop",
+                ));
+            }
+
+            let remaining_hops: Vec<Hop> = hops[1..].to_vec();
+
+            store_route_state(
+                &mut deps.storage,
+                &RouteState {
+                    is_done: false,
+                    remaining_route: Route {
+                        hops: remaining_hops,
+                        to,
+                    },
+                },
+            )?;
+
+            let mut msgs = vec![];
+
+            if let (Some(token_address), Some(token_code_hash)) =
+                (first_hop.from_token.address, first_hop.from_token.code_hash)
+            {
+                // first hop is a snip20
+                msgs.push(snip20::send_msg(
+                    first_hop.pair_address,
+                    amount,
+                    // build swap msg for the next hop
+                    Some(to_binary(&Snip20Swap::Swap {
+                        // set expected_return to None because we don't care about slippage mid-route
+                        expected_return: None,
+                        // set the recepient of the swap to be this contract (the router)
+                        to: Some(env.contract.address.clone()),
+                    })?),
+                    None,
+                    256,
+                    token_code_hash,
+                    token_address,
+                )?);
+            } else if first_hop.from_token.native_denom == Some("uscrt".into()) {
```

Comment:
Why do we have to check if it's "uscrt"? Just send it - it will fail anyway if the asset is not supported